<?xml version="1.0" ?>
<sdf version='1.7'>
  <world name='default'>
    <light name='sun' type='directional'>
      <cast_shadows>1</cast_shadows>
      <pose>0 0 10 0 -0 0</pose>
      <diffuse>0.8 0.8 0.8 1</diffuse>
      <specular>0.2 0.2 0.2 1</specular>
      <attenuation>
        <range>1000</range>
        <constant>0.9</constant>
        <linear>0.01</linear>
        <quadratic>0.001</quadratic>
      </attenuation>
      <direction>-0.5 0.1 -0.9</direction>
      <spot>
        <inner_angle>0</inner_angle>
        <outer_angle>0</outer_angle>
        <falloff>0</falloff>
      </spot>
    </light>
    <model name='ground_plane'>
      <static>1</static>
      <link name='link'>
        <collision name='collision'>
          <geometry>
            <plane>
              <normal>0 0 1</normal>
              <size>100 100</size>
            </plane>
          </geometry>
          <surface>
            <contact>
              <collide_bitmask>65535</collide_bitmask>
              <ode/>
            </contact>
            <friction>
              <ode>
                <mu>100</mu>
                <mu2>50</mu2>
              </ode>
              <torsional>
                <ode/>
              </torsional>
            </friction>
            <bounce/>
          </surface>
          <max_contacts>10</max_contacts>
        </collision>
        <visual name='visual'>
          <cast_shadows>0</cast_shadows>
          <geometry>
            <plane>
              <normal>0 0 1</normal>
              <size>100 100</size>
            </plane>
          </geometry>
          <material>
            <script>
              <uri>file://media/materials/scripts/gazebo.material</uri>
              <name>Gazebo/Grey</name>
            </script>
          </material>
        </visual>
        <self_collide>0</self_collide>
        <enable_wind>0</enable_wind>
        <kinematic>0</kinematic>
      </link>
    </model>
    <gravity>0 0 -9.8</gravity>
    <magnetic_field>6e-06 2.3e-05 -4.2e-05</magnetic_field>
    <atmosphere type='adiabatic'/>
    <physics type='ode'>
      <max_step_size>0.001</max_step_size>
      <real_time_factor>1</real_time_factor>
      <real_time_update_rate>1000</real_time_update_rate>
    </physics>
    <scene>
      <ambient>0.4 0.4 0.4 1</ambient>
      <background>0.7 0.7 0.7 1</background>
      <shadows>1</shadows>
    </scene>
    <wind/>
    <spherical_coordinates>
      <surface_model>EARTH_WGS84</surface_model>
      <latitude_deg>0</latitude_deg>
      <longitude_deg>0</longitude_deg>
      <elevation>0</elevation>
      <heading_deg>0</heading_deg>
    </spherical_coordinates>
    <state world_name='default'>
      <sim_time>68 267000000</sim_time>
      <real_time>68 371195420</real_time>
      <wall_time>1723308750 403390896</wall_time>
      <iterations>68267</iterations>
      <model name='ground_plane'>
        <pose>0 0 0 0 -0 0</pose>
        <scale>1 1 1</scale>
        <link name='link'>
          <pose>0 0 0 0 -0 0</pose>
          <velocity>0 0 0 0 -0 0</velocity>
          <acceleration>0 0 0 0 -0 0</acceleration>
          <wrench>0 0 0 0 -0 0</wrench>
        </link>
      </model>
      <light name='sun'>
        <pose>0 0 10 0 -0 0</pose>
      </light>
    </state>
    <gui fullscreen='0'>
      <camera name='user_camera'>
        <pose>8.70849 -8.57406 24.6049 0 0.999643 2.21619</pose>
        <view_controller>orbit</view_controller>
        <projection_type>perspective</projection_type>
      </camera>
    </gui>

    <!-- Red cube model -->
    <model name="red_cube">
      <pose>3 4 0.5 0 0 0.785</pose> <!-- Cube Position -->
      <static>true</static>
      <link name="link">
        <visual name="visual">
          <geometry>
            <box>
              <size>1 10 1</size> <!-- Cube Size -->
            </box>
          </geometry>
          <material>
            <ambient>1 0 0 1</ambient> <!-- Cube Color -->
          </material>
        </visual>
        <collision name="collision">
          <geometry>
            <box>
              <size>1 1 1</size> <!-- Cube Size -->
            </box>
          </geometry>
        </collision>
      </link>
    </model>

    <!-- Green cube model -->
    <model name="green_cube">
      <pose>-3.5 1.5 0.5 0 0 0</pose> <!-- Cube Position -->
      <static>true</static>
      <link name="link">
        <visual name="visual">
          <geometry>
            <box>
              <size>2 3 1</size> <!-- Cube Size -->
            </box>
          </geometry>
          <material>
            <ambient>0 1 0 1</ambient> <!-- Cube Color -->
          </material>
        </visual>
        <collision name="collision">
          <geometry>
            <box>
              <size>1 1 1</size> <!-- Cube Size -->
            </box>
          </geometry>
        </collision>
      </link>
    </model>

    <!-- Blue cube model -->
    <model name="blue_cube">
      <pose>1 -7 0.5 0 0 0</pose> <!-- Cube Position -->
      <static>true</static>
      <link name="link">
        <visual name="visual">
          <geometry>
            <box>
              <size>5 3 1</size> <!-- Cube Size -->
            </box>
          </geometry>
          <material>
            <ambient>0 0 1 1</ambient> <!-- Cube Color -->
          </material>
        </visual>
        <collision name="collision">
          <geometry>
            <box>
              <size>1 1 1</size> <!-- Cube Size -->
            </box>
          </geometry>
        </collision>
      </link>
    </model>

    <model name="camera_cube">
      <static>true</static>
      <pose>0 0 0.5 0 0 0</pose>
      <link name="base_link">
        <visual name="visual">
          <geometry>
            <box>
              <size>1 1 1</size>
            </box>
          </geometry>
        </visual>
        <sensor name="front_camera" type="camera">
          <pose>0.5 0 0 0 0 0</pose>
          <camera>
            <horizontal_fov>1.047</horizontal_fov>
            <image>
              <width>800</width>
              <height>600</height>
            </image>
          </camera>
          <always_on>1</always_on>
          <update_rate>30</update_rate>
          <visualize>true</visualize>
          <plugin name="camera_controller" filename="libgazebo_ros_camera.so"> <!-- Camerea Image Plugin -->
              <robotNamespace>/</robotNamespace>
              <cameraName>front_camera</cameraName>
              <imageTopicName>image_raw</imageTopicName>
              <cameraInfoTopicName>camera_info</cameraInfoTopicName>
          </plugin>
        </sensor>
        <sensor name="back_camera" type="camera">
          <pose>-0.5 0 0 0 0 3.142</pose>
          <camera>
            <horizontal_fov>1.047</horizontal_fov>
            <image>
              <width>800</width>
              <height>600</height>
            </image>
          </camera>
          <always_on>1</always_on>
          <update_rate>30</update_rate>
          <visualize>true</visualize>
          <plugin name="camera_controller" filename="libgazebo_ros_camera.so"> <!-- Camerea Image Plugin -->
              <robotNamespace>/</robotNamespace>
              <cameraName>back_camera</cameraName>
              <imageTopicName>image_raw</imageTopicName>
              <cameraInfoTopicName>camera_info</cameraInfoTopicName>
          </plugin>
        </sensor>
        <sensor name="left_camera" type="camera">
          <pose>0 0.5 0 0 0 4.712</pose>
          <camera>
            <horizontal_fov>1.047</horizontal_fov>
            <image>
              <width>800</width>
              <height>600</height>
            </image>
          </camera>
          <always_on>1</always_on>
          <update_rate>30</update_rate>
          <visualize>true</visualize>
          <plugin name="camera_controller" filename="libgazebo_ros_camera.so">
              <robotNamespace>/</robotNamespace>
              <cameraName>left_camera</cameraName>
              <imageTopicName>image_raw</imageTopicName>
              <cameraInfoTopicName>camera_info</cameraInfoTopicName>
          </plugin>
        </sensor>
        <sensor name="right_camera" type="camera">
          <pose>0 -0.5 0 0 0 1.571</pose>
          <camera>
            <horizontal_fov>1.047</horizontal_fov>
            <image>
              <width>800</width>
              <height>600</height>
            </image>
          </camera>
          <always_on>1</always_on>
          <update_rate>30</update_rate>
          <visualize>true</visualize>
          <plugin name="camera_controller" filename="libgazebo_ros_camera.so">
              <robotNamespace>/</robotNamespace>
              <cameraName>right_camera</cameraName>
              <imageTopicName>image_raw</imageTopicName>
              <cameraInfoTopicName>camera_info</cameraInfoTopicName>
          </plugin>
        </sensor>
      </link>
      <plugin name="link_pose_control" filename="liblink_pose_control.so"> <!-- Link Pose Control Plugin-->
      </plugin>
    </model>
  </world>
</sdf>
